### ISSUE TITLE
Discussion: Runtime types:  ARC-4 Extension

### ISSUE THREAD

**Issue by SudoWeezy** [algorandfoundation/ARCs#167]
Opening this after the ARC-meeting discussion (15th of February 2023).

Currently, it is not easy to manipulate variable length structures such as strings, dynamic arrays, and complex types at runtime, while being compliant with ARC4. 
We aim to develop a write-efficient way of representing complex structures during runtime, which will reduce the opcode budget required to make changes to these structures.

**Comment by joe-p** [comment]
The main problem we want to solve is that ABI types are very read efficient, which makes sense for wire encoding, but not write efficient. This makes it hard to represent complex structures during runtime in a way that doesn't use too much of the opcode budget when making changes.

My approach involves representing runtime in scratch space, with dynamic types being uint8s pointing to the scratch slot containing the array. The downside with this approach is that unmarshaling can be expensive, but the nice thing is this is only done once. To me, this is much better than potentially expensive writes because only developers that are intimately familiar with ABI encoding will know that dynamic types at the end of a tuple are more expensive to write than those in the beginning. 

The main thing I still need to figure out is how to handle large complex data structures in box storage. I'm still not convinced this is the best approach, but it's what I've come up with so far.

Below is an example with dynamic arrays, more examples can be seen here: https://github.com/joe-p/TEALScript/blob/b08529d428a9cc9d721640b2241a01cca9a0f4b1/tests/contracts/abi.ts


# TEALScript
```typescript
  threeDimensionalArray(): void {
    const c: uint64[][][] = [[[11, 22], [33, 44]], [[55, 66], [77, 88]]];

    assert(c[1][1][1] === 88);
  }
```

# Scratch Representation
At the end of unmarshaling this is what the scratch slots look like

Slot | Value | Comment
---  | --- | ---
0 | `8` | Scratch slot 0 stores the next scratch slot to use for the next dynamic type
1 | `0x000000000000000b0000000000000016` | `[11, 22]`
2 | `0x0000000000000021000000000000002c` | `[33, 44]`
3 | `0x0102` | Pointers to scratch slots 1 and 2
4 | `0x00000000000000370000000000000042` | `[55, 66]`
5 | `0x000000000000004d0000000000000058` | `[77, 88]`
6 | `0x0405` | Pointers to scratch slots 4 and 5
7 | `0x0306` | Pointer to scratch slots 3 and 6

# TEAL

```
  pc# |ln# |source         |scratch                                |stack
    1   |1   |intcblock 1 88 |                                       |[]
    5   |2   |bytecblock 0x  |                                       |[]
    8   |3   |intc_0 // 1    |                                       |[]
    9   |4   |store 0        |                                       |[1]
    11  |5   |b label1       |0 = 1                                  |[]
    37  |23  |pushint 11     |                                       |[]
    39  |24  |itob           |                                       |[11]
    40  |25  |pushint 22     |                                       |[0x000000000000000b]
    42  |26  |itob           |                                       |[0x000000000000000b, 22]
    43  |27  |bytec_0 // ""  |                                       |[0x000000000000000b, 0x0000000000000016]
    44  |28  |concat         |                                       |[0x000000000000000b, 0x0000000000000016, 0x]
    45  |29  |concat         |                                       |[0x000000000000000b, 0x0000000000000016]
    46  |30  |callsub label2 |                                       |[0x000000000000000b0000000000000016]
    14  |7   |load 0         |                                       |[0x000000000000000b0000000000000016]
    16  |8   |swap           |                                       |[0x000000000000000b0000000000000016, 1]
    17  |9   |stores         |                                       |[1, 0x000000000000000b0000000000000016]
    18  |10  |load 0         |1 = 0x000000000000000b0000000000000016 |[]
    20  |11  |itob           |                                       |[1]
    21  |12  |extract 7 1    |                                       |[0x0000000000000001]
    24  |13  |load 0         |                                       |[0x01]
    26  |14  |intc_0 // 1    |                                       |[0x01, 1]
    27  |15  |+              |                                       |[0x01, 1, 1]
    28  |16  |store 0        |                                       |[0x01, 2]
    30  |17  |retsub         |0 = 2                                  |[0x01]
    49  |31  |pushint 33     |                                       |[0x01]
    51  |32  |itob           |                                       |[0x01, 33]
    52  |33  |pushint 44     |                                       |[0x01, 0x0000000000000021]
    54  |34  |itob           |                                       |[0x01, 0x0000000000000021, 44]
    55  |35  |bytec_0 // ""  |                                       |[0x01, 0x0000000000000021, 0x000000000000002c]
    56  |36  |concat         |                                       |[0x01, 0x0000000000000021, 0x000000000000002c, 0x]
    57  |37  |concat         |                                       |[0x01, 0x0000000000000021, 0x000000000000002c]
    58  |38  |callsub label2 |                                       |[0x01, 0x0000000000000021000000000000002c]
    14  |7   |load 0         |                                       |[0x01, 0x0000000000000021000000000000002c]
    16  |8   |swap           |                                       |[0x01, 0x0000000000000021000000000000002c, 2]
    17  |9   |stores         |                                       |[0x01, 2, 0x0000000000000021000000000000002c]
    18  |10  |load 0         |2 = 0x0000000000000021000000000000002c |[0x01]
    20  |11  |itob           |                                       |[0x01, 2]
    21  |12  |extract 7 1    |                                       |[0x01, 0x0000000000000002]
    24  |13  |load 0         |                                       |[0x01, 0x02]
    26  |14  |intc_0 // 1    |                                       |[0x01, 0x02, 2]
    27  |15  |+              |                                       |[0x01, 0x02, 2, 1]
    28  |16  |store 0        |                                       |[0x01, 0x02, 3]
    30  |17  |retsub         |0 = 3                                  |[0x01, 0x02]
    61  |39  |bytec_0 // ""  |                                       |[0x01, 0x02]
    62  |40  |concat         |                                       |[0x01, 0x02, 0x]
    63  |41  |concat         |                                       |[0x01, 0x02]
    64  |42  |callsub label2 |                                       |[0x0102]
    14  |7   |load 0         |                                       |[0x0102]
    16  |8   |swap           |                                       |[0x0102, 3]
    17  |9   |stores         |                                       |[3, 0x0102]
    18  |10  |load 0         |3 = 0x0102                             |[]
    20  |11  |itob           |                                       |[3]
    21  |12  |extract 7 1    |                                       |[0x0000000000000003]
    24  |13  |load 0         |                                       |[0x03]
    26  |14  |intc_0 // 1    |                                       |[0x03, 3]
    27  |15  |+              |                                       |[0x03, 3, 1]
    28  |16  |store 0        |                                       |[0x03, 4]
    30  |17  |retsub         |0 = 4                                  |[0x03]
    67  |43  |pushint 55     |                                       |[0x03]
    69  |44  |itob           |                                       |[0x03, 55]
    70  |45  |pushint 66     |                                       |[0x03, 0x0000000000000037]
    72  |46  |itob           |                                       |[0x03, 0x0000000000000037, 66]
    73  |47  |bytec_0 // ""  |                                       |[0x03, 0x0000000000000037, 0x0000000000000042]
    74  |48  |concat         |                                       |[0x03, 0x0000000000000037, 0x0000000000000042, 0x]
    75  |49  |concat         |                                       |[0x03, 0x0000000000000037, 0x0000000000000042]
    76  |50  |callsub label2 |                                       |[0x03, 0x00000000000000370000000000000042]
    14  |7   |load 0         |                                       |[0x03, 0x00000000000000370000000000000042]
    16  |8   |swap           |                                       |[0x03, 0x00000000000000370000000000000042, 4]
    17  |9   |stores         |                                       |[0x03, 4, 0x00000000000000370000000000000042]
    18  |10  |load 0         |4 = 0x00000000000000370000000000000042 |[0x03]
    20  |11  |itob           |                                       |[0x03, 4]
    21  |12  |extract 7 1    |                                       |[0x03, 0x0000000000000004]
    24  |13  |load 0         |                                       |[0x03, 0x04]
    26  |14  |intc_0 // 1    |                                       |[0x03, 0x04, 4]
    27  |15  |+              |                                       |[0x03, 0x04, 4, 1]
    28  |16  |store 0        |                                       |[0x03, 0x04, 5]
    30  |17  |retsub         |0 = 5                                  |[0x03, 0x04]
    79  |51  |pushint 77     |                                       |[0x03, 0x04]
    81  |52  |itob           |                                       |[0x03, 0x04, 77]
    82  |53  |intc_1 // 88   |                                       |[0x03, 0x04, 0x000000000000004d]
    83  |54  |itob           |                                       |[0x03, 0x04, 0x000000000000004d, 88]
    84  |55  |bytec_0 // ""  |                                       |[0x03, 0x04, 0x000000000000004d, 0x0000000000000058]
    85  |56  |concat         |                                       |[0x03, 0x04, 0x000000000000004d, 0x0000000000000058, 0x]
    86  |57  |concat         |                                       |[0x03, 0x04, 0x000000000000004d, 0x0000000000000058]
    87  |58  |callsub label2 |                                       |[0x03, 0x04, 0x000000000000004d0000000000000058]
    14  |7   |load 0         |                                       |[0x03, 0x04, 0x000000000000004d0000000000000058]
    16  |8   |swap           |                                       |[0x03, 0x04, 0x000000000000004d0000000000000058, 5]
    17  |9   |stores         |                                       |[0x03, 0x04, 5, 0x000000000000004d0000000000000058]
    18  |10  |load 0         |5 = 0x000000000000004d0000000000000058 |[0x03, 0x04]
    20  |11  |itob           |                                       |[0x03, 0x04, 5]
    21  |12  |extract 7 1    |                                       |[0x03, 0x04, 0x0000000000000005]
    24  |13  |load 0         |                                       |[0x03, 0x04, 0x05]
    26  |14  |intc_0 // 1    |                                       |[0x03, 0x04, 0x05, 5]
    27  |15  |+              |                                       |[0x03, 0x04, 0x05, 5, 1]
    28  |16  |store 0        |                                       |[0x03, 0x04, 0x05, 6]
    30  |17  |retsub         |0 = 6                                  |[0x03, 0x04, 0x05]
    90  |59  |bytec_0 // ""  |                                       |[0x03, 0x04, 0x05]
    91  |60  |concat         |                                       |[0x03, 0x04, 0x05, 0x]
    92  |61  |concat         |                                       |[0x03, 0x04, 0x05]
    93  |62  |callsub label2 |                                       |[0x03, 0x0405]
    14  |7   |load 0         |                                       |[0x03, 0x0405]
    16  |8   |swap           |                                       |[0x03, 0x0405, 6]
    17  |9   |stores         |                                       |[0x03, 6, 0x0405]
    18  |10  |load 0         |6 = 0x0405                             |[0x03]
    20  |11  |itob           |                                       |[0x03, 6]
    21  |12  |extract 7 1    |                                       |[0x03, 0x0000000000000006]
    24  |13  |load 0         |                                       |[0x03, 0x06]
    26  |14  |intc_0 // 1    |                                       |[0x03, 0x06, 6]
    27  |15  |+              |                                       |[0x03, 0x06, 6, 1]
    28  |16  |store 0        |                                       |[0x03, 0x06, 7]
    30  |17  |retsub         |0 = 7                                  |[0x03, 0x06]
    96  |63  |bytec_0 // ""  |                                       |[0x03, 0x06]
    97  |64  |concat         |                                       |[0x03, 0x06, 0x]
    98  |65  |concat         |                                       |[0x03, 0x06]
    99  |66  |callsub label2 |                                       |[0x0306]
    14  |7   |load 0         |                                       |[0x0306]
    16  |8   |swap           |                                       |[0x0306, 7]
    17  |9   |stores         |                                       |[7, 0x0306]
    18  |10  |load 0         |7 = 0x0306                             |[]
    20  |11  |itob           |                                       |[7]
    21  |12  |extract 7 1    |                                       |[0x0000000000000007]
    24  |13  |load 0         |                                       |[0x07]
    26  |14  |intc_0 // 1    |                                       |[0x07, 7]
    27  |15  |+              |                                       |[0x07, 7, 1]
    28  |16  |store 0        |                                       |[0x07, 8]
    30  |17  |retsub         |0 = 8                                  |[0x07]
    102 |67  |btoi           |                                       |[0x07]
    103 |68  |loads          |                                       |[7]
    104 |69  |extract 1 1    |                                       |[0x0306]
    107 |70  |btoi           |                                       |[0x06]
    108 |71  |loads          |                                       |[6]
    109 |72  |extract 1 1    |                                       |[0x0405]
    112 |73  |btoi           |                                       |[0x05]
    113 |74  |loads          |                                       |[5]
    114 |75  |extract 8 8    |                                       |[0x000000000000004d0000000000000058]
    117 |76  |btoi           |                                       |[0x0000000000000058]
    118 |77  |intc_1 // 88   |                                       |[88]
    119 |78  |==             |                                       |[88, 88]
    120 |79  |assert         |                                       |[1]
    121 |80  |intc_0 // 1    |                                       |[]
    122 |81  |return         |                                       |[1]
    123 |82  |               |                                       |[1]
```

**Comment by FrankSzendzielarz** [comment]
For posterity, some thoughts of my own from prior conversations on Discord:

For example with structures like this :

```csharp
struct Player{
     ushort Id;
     Weapon[] Weapons;
}

struct Weapon{
    string Name;
    ulong Strength;
}

```
where the ARC4 encoding would be something like this

``(uint16, (string,uint64)[])``

Then with the existing read-optimised encoding described in ARC4 :

if we wanted to compile the syntax 

```csharp
      player.Weapons[2].Name
```
runtime get/set could be facilitated by opcode syntaxes loosely like this

``` get offset 2 dynamic[2] offset 0 dynamic ```

In general to support mutable variable length information like arrays and string, there are only two possible strategies that I can see:

1. Change the encoding to something that is friendly both to read and update
2. Introduce AVM new opcodes [and potentially types ] that abstract away whatever the wire encoding is.

Using TEAL as @joe-p suggests above is a partial solution to strategy 2. But still there are requirements such as calling inner app call transactions that would require re-encoding and decoding within a transaction, giving additional overhead, that would be removed if there was an AVM native representation of complex types and/or opcodes for complex type accessors.

My approach is that for the C# tooling we'll be using a different encoding for complex types than specified in ARC4 until this issue is resolved.

**Comment by FrankSzendzielarz** [comment]
> Opening this after the ARC-meeting discussion (15th of February 2023).

We need to add what the issue is about. Also I don't see this as an extension to ARC4.

This is about making it convenient, or even possible, to manipulate variable length structures like strings, dynamic arrays and complex types at runtime, if you are compliant with ARC4. In essence it represents a decision to abstract away the wire encoding and come up with some way of figuring out how to represent complex types / arrays in TEAL.

**Comment by joe-p** [comment]
> My approach is that for the C# tooling we'll be using a different encoding for complex types than specified in ARC4 until this issue is resolved.

I'd really like to see how you are approaching this written down. From my understanding, your plan is to represent tuples simply by concatenating all of the types. For example, `[[1, 2], [3, 4]] // (uint8[], uint16[])` would be `0x 0002 01 02 0002 0003 0004`. To read you would have to extract `len(x)`, `btoi` it, then add to total offset and repeat. This needs to be done for every single read AND every write (so you know where to split the full data structure for insertion). If you read before writing, you could save that offset in scratch or on the stack, but then any intermediate changes to proceeding elements in the tuple would require you to update all of your saved offsets.

**Comment by FrankSzendzielarz** [comment]
> 

yes , for 'variable' types the length is prefixed to each structure element. to get to the correct element you need to iteratively ``extract_uint16`` to get to the structure element you want.  However, an update to the structure would then be just concatenating the preceding bytes, the new element, and the trailing bytes.

**Comment by joe-p** [comment]
> but with the assumption that structures will not have that many fields

I think this is the main problem for both your approach and my approach. I believe I need to go back to the drawing board and I think it'd be wise if we still put some more thought into this. With box storage and the limitation on box references, it's plausibly desirable for a single box to hold a complex data type with many fields.  

> Also, I am still scratching my head about how, in the current ARC4, you calculate the modified length of tuples of dynamic types: I am not sure one can trust the head list as a source of element length or should. But I don't really have time to think about all that anymore.

If the tuple is encoded in compliance with ARC4 the head list can be trusted. The head encoding for `x[i]` is specified as `head(x[i]) = enc(len( head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[i-1]) ))` which is not ambiguous. It's true that this is simply the offset of the dynamic type and assuming such "works", but ARC4-compliant encoding will always be as small as possible.

**Comment by joe-p** [comment]
I believe I have came up with an approach that could potentially be `O(1)` for both read and writes for tuples. For this to truly be `O(1)` we'd need an additional opcode `repeat n` that functions like golang's `byte.Repeat()` which I am guessing is much more reasonable to implement than direct support for complex data types.

It would also require all static types in a tuple to be 0-padded to ensure for every `x` `bytelen(head(x))` is divisible by 2 (worst case one extra byte per static type). This could be done in a update to the ABI or as an additional encoding/decoding step during runtime. I think it's unreasonable to amend ARC-0004 now, but with AVM9 we will need a new ABI standard that is backwards incompatible and I would propose that this is implemented then. 

The write algorithm:

```
// Code that uses extract/concat to add new data to the tuple goes here (which is O(1))

// Now for the head update...

// Let this be the current head for (byte[], uint32, byte[])
// 0x0008_12345678_3210
byte 0x0008123456783210

// Let's say we're increasing the len of the first element by 3
byte 0x0003

// Repeat this value to match the byte length of the head
// Repeat opcode would replace this "dupn; concat..." pattern, otherwise the concats make it O(N)
dupn 3
concat
concat
concat

// Let this be a head update bitmask (from compiler) that signifies which heads are getting updated
// not-updated_uint32_updated
// 0x0000_00000000_FFFF
byte 0x000000000000FFFF 
b&
b+ // TODO: some logic to determine if add or subtract

// replace existing head with the value on the stack here
```

This algo is somewhat expensive, but the important thing here is that it's `O(1)` (assuming we can get a `repeat n` opcode and an ABI update) which is very important

**Comment by FrankSzendzielarz** [comment]
That's nice!

There are still issues with using the ABI encoded type directly as it is. For example, arrays of bools would be simplified if they were just lists of bytes rather than flags. Or calculating the change in size for the reencoded element and then "recursively" applying the changes to nested elements. 

I think someone then needs to work out what the effort would be or risk would be in just adding opcodes that work with an encoding, like the other idea above:

``get offset 2 dynamic[2] offset 0 dynamic``

vs tweaking the encoding to be convenient for TEAL and tweaking TEAL slightly.

**Comment by joe-p** [comment]
After talking with JJ on discord I've realized a fatal flaw, `b+` and `b-` only work with 64 bytes so it's not O(1). There still might be value had in using an algo like this, so I'll play around with it

**Comment by joe-p** [comment]
After more discussion and thought I think the best approach is enforcing static types in storage. The obvious downside is the potential for space inefficiency, but storage space is being paid for by MBR so I don't think this is a significant reason to not implement. Space efficient read-optimized encoding for ABI and read/write optimized encoding for internal (storage and runtime) types makes the most sense to me.

**Comment by FrankSzendzielarz** [comment]
> n using an algo like this, so I'll

Do you reckon? I was thinking about that also and came to the conclusion that it's far too restrictive. A wire encoded 4096 byte array will decode into what?  You might as well just use a standard compression algorithm on the runtime encoding and avoid the arc4 compressions.

For example: you have a Player that has an Inventory of Magic Items , Shield and Weapons. You cannot carry more than 10 of each and no more than 10 items in total. With static representation you always have 30 items, which may exceed 4096 bytes for example, but with a dynamic representation it's always under 4096 bytes and never more than 10 items.

Still, I am not clear why we would then want any kind of dynamic representation in the wire encoding (and not say, a standard gzip type thing instead)

**Comment by FrankSzendzielarz** [comment]
> After talking with JJ on discord I've realized a fatal flaw, `b+` and `b-` only work with 64 bytes so it's not O(1). There still might be value had in using an algo like this, so I'll play around with it

I don't understand why that matters. The opcode cost is the same and the output result is always truncated to the byte length you need.

**Comment by joe-p** [comment]
> You might as well just use a standard compression algorithm on the runtime encoding and avoid the arc4 compressions.

But this would surely be more computational expensive to decompress/decode, no?

> For example: you have a Player that has an Inventory of Magic Items , Shield and Weapons. You cannot carry more than 10 of each and no more than 10 items in total. With static representation you always have 30 items, which may exceed 4096 bytes for example, but with a dynamic representation it's always under 4096 bytes and never more than 10 items.

I see what you are saying, but for this specific scenario I'd really question why you'd be storing that much data directly on-chain. Pointing to an IPFS hash makes way more sense. Ignoring that, you could have all the items be the same data structure with a `type` field. Another option is having the different items stored in different boxes. There's also no reason the compiler couldn't support iterating over pages of 4kb data, but I understand this is much less space efficient. 

> I don't understand why that matters. The opcode cost is the same and the output result is always truncated to the byte length you need.

If the head is longer than 64 bytes, you'd need to do the repeat step multiple times which means it's no longer O(1) with respect to opcode cost.

**Comment by FrankSzendzielarz** [comment]
> But this would surely be more computational expensive to decompress/decode, no?

I mean, have the AVM do the grunt work for you. 

Alternatively not use compression, but then the message sizes in the txns will be a bit larger.

> I see what you are saying, but for this specific scenario I'd really question why you'd be storing that much data directly on-chain.

Never mind the specifics...this is just some 4096 byte array being sent to a smart contract, or returned from - 4k isn't that much really. On the other hand it might be useful to just set the hard limitation (everything will be fixed size) and see how people respond. It feels weird to me. Forcing every string to be set to "max" length in a structure. But it might be enough, given the possible use cases. Probably worth doing a bit of market research, so to speak.

> If the head is longer than 64 bytes, you'd need to do the repeat step multiple times which means it's no longer O(1) with respect to opcode cost.

Oh ok I see. But in the context above we were talking about tweaking TEAL instructions so a 4096 byte bitmask operator could be introduced :-)

**Comment by joe-p** [comment]
> I mean, have the AVM do the grunt work for you.

Whether it's done automatically or not by the AVM is somewhat irrelevant. Yes it would be more convenient for devs, but the main thing we care about is the total time it takes for algod to eval a transaction. I imagine a per-txn decompression would significantly slow down the total eval time. 

> 4k isn't that much really.

4k is A LOT in the context of blockchains. Given our current data cost model it doesn't seem like a lot, but I'm not convinced it's sustainable. 4kb on ETH for example is in the millions of dollars. I don't think that's the proper cost model either, but in general if we should store as much data as possible on IPFS and only save things that are actually used on-chain in on-chain storage. 

> Oh ok I see. But in the context above we were talking about tweaking TEAL instructions so a 4096 byte bitmask operator could be introduced :-)

Hahaha yeah unfortunately if we did have a more robust byte arithmetic the cost would vary depending on the size of the inputs so still not O(1) unfortunately.

**Comment by FrankSzendzielarz** [comment]
> Whether it's done automatically or not by the AVM is somewhat irrelevant. 

I know what you mean. The question is really for the protocol designers then.... It seems that by having an inconvenient encoding with bitflags representing boolean areas and so on, then we have the compute cost passed onto the developer...with the additional minus of it being a PITA which reduces adoption.

But if the encode/decode was done by the AVM then that minus of complexity/adoption is removed but the protocol developers can choose a more optimal encoding. Who knows, the alternative of passing uncompressed large fixed size stuff around might involve more compute than compressed.

**Comment by FrankSzendzielarz** [comment]
Actually an afterthought. The inner transaction scenario...ideally we have an agnostic way of calling ABI methods so that you avoid the encode/decode step for inner transactions when we pass modified structures through.

**Comment by joe-p** [comment]
Shared on discord but sharing here as well for posterity sake. 

There are two main implementations being discussed:

1. Following ARC4 encoding and use a bitmask and byte arithmetic to update all heads at once 
2. Encode arrays by simply putting them in place sequentially with all of them being prefixed by their byte length (not element length)

I will refer to 1 as the TEALScript implementation (since that is what I'm doing for TEALScript) and 2 as the simple implementation. 

The current TEALScript implementation consumes 80-95 ops to append a value to a dynamic array in a tuple. Reading or updating a value is 15-20 ops. The limitations of the implementation is that the dynamic arrays must come at the end of the tuple and it doesn't allow nested dynamic types. Both of these could be solved at the compiler/language level though. 

For the simple implementation, reading, writing or updating a value in an array would roughly cost `i * 8`  ops with `i` being the index of the dynamic array in the tuple, assuming `i` is `0` for the first dynamic array and all subsequent elements are dynamic arrays. In other words, this is assuming all dynamic arrays come at the end of a tuple, much like above, but isn't a strict requirement (although not doing so would make for inefficient reads of static types as well). 

The full TEALScript implementation can be seen here: https://github.com/joe-p/TEALScript/blob/93696e6181ce88ee24b64ce7b16ffa35d366ad0f/tests/contracts/AbiTest.approval.teal#L1400-L1462
The simple implementation can be seen here: https://gist.github.com/joe-p/458d8d591df67091244e2b2f4d41872d

TL;DR 
* Simple implementation is more read efficient up to ~2 dynamic arrays, after which TEALScript is more efficient
* Simple implementation is more write efficient up to ~10 dynamic arrays, after which TEALScript is more efficient
* TEALScript implementation is limited to 32 dynamic arrays in a single tuple
